docker/docker_image_dest.go

```diff
@@ -131,15 +132,49 @@ func (d *dockerImageDestination) HasThreadSafePutBlob() bool {
 // to any other readers for download using the supplied digest.
 // If stream.Read() at any time, ESPECIALLY at end of input, returns an error, PutBlob MUST 1) fail, and 2) delete any data stored so far.
 func (d *dockerImageDestination) PutBlob(ctx context.Context, stream io.Reader, inputInfo types.BlobInfo, cache types.BlobInfoCache, isConfig bool) (types.BlobInfo, error) {
+	digester, stream := putblobdigest.DigestIfCanonicalUnknown(stream, inputInfo)
+	sizeCounter := &sizeCounter{}
+	stream = io.TeeReader(stream, sizeCounter)
+
+	// If requested, precompute the blob digest to prevent uploading layers that already exist on the registry.
+	// This functionality is particularly useful when BlobInfoCache has not been populated with compressed digests,
+	// the source blob is uncompressed, and the destination blob is being compressed "on the fly".
+	if inputInfo.Digest == "" && d.c.sys.DockerRegistryPushPrecomputeDigests {
+		logrus.Debugf("Precomputing digest layer for %s", reference.Path(d.ref.ref))
+		diskBlob, err := ioutil.TempFile(tmpdir.TemporaryDirectoryForBigFiles(d.c.sys), "docker-layer-blob")
+		if err != nil {
+			return types.BlobInfo{}, errors.Wrap(err, "creating temporary on-disk layer")
+		}
+		defer diskBlob.Close()
+		// immediately unlink for removal on fd close
+		err = os.Remove(diskBlob.Name())
+		if err != nil {
+			return types.BlobInfo{}, errors.Wrap(err, "unlinking temporary on-disk layer")
+		}
+		_, err = io.Copy(diskBlob, stream)
+		if err != nil {
+			return types.BlobInfo{}, errors.Wrap(err, "writing to temporary on-disk layer")
+		}
+		_, err = diskBlob.Seek(0, io.SeekStart)
+		if err != nil {
+			return types.BlobInfo{}, errors.Wrap(err, "rewinding temporary on-disk layer")
+		}
+		stream = diskBlob
+		inputInfo.Digest = digester.Digest()
+		inputInfo.Size = sizeCounter.size
+	}
+
 	if inputInfo.Digest != "" {
-		// This should not really be necessary, at least the copy code calls TryReusingBlob automatically.
-		// Still, we need to check, if only because the "initiate upload" endpoint does not have a documented "blob already exists" return value.
+		// The copy code calls TryReusingBlob automatically; however, we still need to perform an existing layer check for only this blob.
```

Comment:
The copy code is not the _only_ anticipated caller of this method (`types.ImageDestination` is, for better or worse, a public type), so the comment should not be changed to assume only that caller.

Or if the change is trying to say something else non-obvious, I have missed the intent of that.

---

… is this trying to say that we have to try _again_ because we didn’t know the digest and now we maybe do because we computed it ourselves? It’s not actually saying that.